cpp/src/link_analysis/hits_impl.cuh

```diff
@@ -0,0 +1,201 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <thrust/fill.h>
+#include <thrust/transform.h>
+#include <cugraph/algorithms.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/prims/copy_to_adj_matrix_row_col.cuh>
+#include <cugraph/prims/copy_v_transform_reduce_in_out_nbr.cuh>
+#include <cugraph/prims/count_if_v.cuh>
+#include <cugraph/prims/reduce_v.cuh>
+#include <cugraph/prims/row_col_properties.cuh>
+#include <cugraph/prims/transform_reduce_v.cuh>
+
+// TODO : delete
+#include <string>
+
+namespace cugraph {
+namespace detail {
+template <typename GraphViewType, typename result_t>
+void normalize(raft::handle_t const& handle,
+               GraphViewType const& graph_view,
+               result_t* hubs,
+               raft::comms::op_t op)
+{
+  auto hubs_norm = reduce_v(handle,
+                            graph_view,
+                            hubs,
+                            hubs + graph_view.get_number_of_local_vertices(),
+                            identity<result_t>(op),
+                            op);
+  thrust::transform(handle.get_thrust_policy(),
+                    hubs,
+                    hubs + graph_view.get_number_of_local_vertices(),
+                    thrust::make_constant_iterator(hubs_norm),
+                    hubs,
+                    thrust::divides<result_t>());
+}
+
+template <typename GraphViewType, typename result_t>
+std::tuple<result_t, size_t> hits(raft::handle_t const& handle,
+                                  GraphViewType const& graph_view,
+                                  result_t* const hubs,
+                                  result_t* const authorities,
+                                  result_t epsilon,
+                                  size_t max_iterations,
+                                  bool has_initial_hubs_guess,
+                                  bool normalize,
+                                  bool do_expensive_check)
+{
+  using vertex_t = typename GraphViewType::vertex_type;
+  static_assert(std::is_integral<vertex_t>::value,
+                "GraphViewType::vertex_type should be integral.");
+  static_assert(std::is_floating_point<result_t>::value,
+                "result_t should be a floating-point type.");
+  static_assert(GraphViewType::is_adj_matrix_transposed,
+                "GraphViewType should support the pull model.");
+
+  auto const num_vertices = graph_view.get_number_of_vertices();
+  result_t hubs_error{std::numeric_limits<result_t>::max()};
```

Comment:
I am a bit uncomfortable with this name. ```hubs_error``` sounds like something is wrong. This isn't really an error but very natural in many numerical methods with iterative solvers... I will also think about a better name.

Reply:
The term residual is used in numerical algorithms for calculating a stopping point.

Reply:
I initially thought users will see this name, but this is used only inside the implementation, and I am less concerned about this name now. I may still use something like ```diff_sum``` instead, but I will stop complaining :-)

Reply:
```residual``` means in solving Ax=b, residual = b - Ax_approx.

Here, ```hubs_error``` is used for sum(diff(cur_hubs[] - prev_hubs[])). I guess these two are not same.